core-java-modules/core-java-io-2/src/test/java/com/baeldung/filenotfoundexception/FileNotFoundExceptionUnitTest.java

```diff
@@ -1,58 +1,116 @@
+    }
+
+    @Test
+    void whenFileExistButInaccessible_thenFileNotFoundException() {
+        File readOnlyFile = tempDir.resolve("dummy.ReadOnly")
+            .toFile();
+        // first creating a read-only file
+        try (PrintWriter printWriter = new PrintWriter(new FileWriter(readOnlyFile))) {
+            printWriter.println("Hi there!");
+            readOnlyFile.setReadOnly();
+        } catch (Exception e) {
+            fail();
         }
+        assertTrue(readOnlyFile.exists());
+        assertFalse(readOnlyFile.canWrite());
+
+        // now we create a FileWriter on the read-only file
+        assertThrows(FileNotFoundException.class, () -> {
+            FileWriter fileWriter = new FileWriter(readOnlyFile);
+        });
     }
 
     @Test
-    public void createFile() throws IOException {
+    void whenFileNotFoundException_thenLogWarn() {
+        Path path = tempDir.resolve("dummy.notExist");
         try {
-            readFailingFile();
+            FileReader fileReader = new FileReader(path.toFile());
         } catch (FileNotFoundException ex) {
-            try {
-                new File(fileName).createNewFile();
-                readFailingFile();
-            } catch (IOException ioe) {
-                throw new RuntimeException("BusinessException: even creation is not possible.");
-            }
+            LOG.warn("Error Occurred when reading the optional file " + path, ex);
         }
     }
 
-    @Test
-    public void logError() throws IOException {
-        try {
-            readFailingFile();
+    public String readConfig(Path path) throws IOException {
+        try (BufferedReader reader = new BufferedReader(new FileReader(path.toFile()))) {
+            return reader.readLine();
         } catch (FileNotFoundException ex) {
-            LOG.error("Optional file " + fileName + " was not found.");
+            throw new ConfigurationException("Reading configuration file failed", ex);
         }
     }
 
-    private void readFailingFile() throws IOException {
-        BufferedReader rd = new BufferedReader(new FileReader(new File(fileName)));
-        rd.readLine();
-        // no need to close file
+    @Test
+    public void whenFileNotExist_thenGetBusinessException() {
+        assertThrows(ConfigurationException.class, () -> readConfig(tempDir.resolve("dummy.notExist")));
     }
 
-    private class BusinessException extends RuntimeException {
-        BusinessException(String string) {
-            super(string);
+    public String readOrCreateFileWithContent(Path path, String defaultValue) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(path.toFile()))) {
+            return reader.readLine();
+        } catch (FileNotFoundException e) {
+            try (PrintWriter writer = new PrintWriter(new FileWriter(path.toFile()))) {
+                writer.print(defaultValue);
+                writer.close();
```

Comment:
This should be implicit as part of the try-with-resources?

Reply:
removed. please check the new commit